File: tui/src/render/markdown.ts
/**
 * Simple streaming-aware markdown renderer for terminal output.
 *
 * Renders basic markdown formatting with ANSI codes:
 * - **bold** and __bold__
 * - *italic* and _italic_
 * - `inline code`
 * - # Headers
 * - - Bullet lists
 * - Code blocks (```lang ... ```)
 */

import pc from 'picocolors';

type RenderState =
  | { mode: 'text' }
  | { mode: 'codeblock'; lang: string; buffer: string };

/**
 * Streaming markdown renderer
 */
export class MarkdownRenderer {
  private state: RenderState = { mode: 'text' };
  private buffer = '';

  /**
   * Push a text delta and get any immediately renderable content
   */
  push(delta: string): string {
    this.buffer += delta;
    return this.processBuffer();
  }

  /**
   * Flush any remaining buffered content
   */
  flush(): string {
    const result = this.renderText(this.buffer);
    this.buffer = '';
    this.state = { mode: 'text' };
    return result;
  }

  /**
   * Reset state for a new message
   */
  reset(): void {
    this.buffer = '';
    this.state = { mode: 'text' };
  }

  /**
   * Process the buffer and return renderable content
   */
  private processBuffer(): string {
    // Check for code block transitions
    if (this.state.mode === 'text') {
      const codeBlockStart = this.buffer.match(/```(\w*)\n?/);
      if (codeBlockStart) {
        const beforeBlock = this.buffer.slice(0, codeBlockStart.index);
        const afterStart = this.buffer.slice(
          (codeBlockStart.index || 0) + codeBlockStart[0].length
        );

        // Render text before code block
        const rendered = this.renderText(beforeBlock);

        // Enter code block mode
        this.state = { mode: 'codeblock', lang: codeBlockStart[1] || '', buffer: '' };
        this.buffer = afterStart;

        // Check if code block ends in same delta
        return rendered + this.processBuffer();
      }

      // If we have complete lines, render them
      const lastNewline = this.buffer.lastIndexOf('\n');
      if (lastNewline >= 0) {
        const complete = this.buffer.slice(0, lastNewline + 1);
        this.buffer = this.buffer.slice(lastNewline + 1);
        return this.renderText(complete);
      }

      return '';
    }

    // In code block mode, look for closing ```
    if (this.state.mode === 'codeblock') {
      const closeMatch = this.buffer.match(/\n?```/);
      if (closeMatch) {
        const codeContent = this.buffer.slice(0, closeMatch.index);
        const afterBlock = this.buffer.slice(
          (closeMatch.index || 0) + closeMatch[0].length
        );

        // Render code block
        const rendered = this.renderCodeBlock(
          this.state.lang,
          this.state.buffer + codeContent
        );

        // Exit code block mode
        this.state = { mode: 'text' };
        this.buffer = afterBlock;

        return rendered + this.processBuffer();
      }

      // Buffer more code content
      this.state.buffer += this.buffer;
      this.buffer = '';
      return '';
    }

    return '';
  }

  /**
   * Render inline markdown formatting
   */
  private renderText(text: string): string {
    if (!text) return '';

    let result = text;

    // Headers (only at line start)
    result = result.replace(/^### (.+)$/gm, pc.bold('$1'));
    result = result.replace(/^## (.+)$/gm, pc.bold(pc.blue('$1')));
    result = result.replace(/^# (.+)$/gm, pc.bold(pc.cyan('$1')));

    // Bold (** or __)
    result = result.replace(/\*\*(.+?)\*\*/g, pc.bold('$1'));
    result = result.replace(/__(.+?)__/g, pc.bold('$1'));

    // Italic (* or _) - but not inside words
    result = result.replace(/(?<![a-zA-Z])\*([^*]+)\*(?![a-zA-Z])/g, pc.italic('$1'));
    result = result.replace(/(?<![a-zA-Z])_([^_]+)_(?![a-zA-Z])/g, pc.italic('$1'));

    // Inline code
    result = result.replace(/`([^`]+)`/g, pc.cyan('$1'));

    // Bullet lists
    result = result.replace(/^(\s*)[*-] /gm, '$1• ');

    // Links [text](url)
    result = result.replace(/\[([^\]]+)\]\(([^)]+)\)/g, `$1 ${pc.dim('($2)')}`);

    return result;
  }

  /**
   * Render a code block with optional syntax highlighting
   */
  private renderCodeBlock(lang: string, code: string): string {
    const trimmedCode = code.trim();
    if (!trimmedCode) return '';

    const lines = trimmedCode.split('\n');
    const langLabel = lang ? pc.dim(`[${lang}]`) : '';
    const border = pc.dim('│');

    let result = '\n';
    if (langLabel) {
      result += `  ${langLabel}\n`;
    }

    for (const line of lines) {
      // Basic syntax highlighting based on common patterns
      let highlighted = line;

      if (lang === 'typescript' || lang === 'ts' || lang === 'javascript' || lang === 'js') {
        highlighted = highlightJS(line);
      } else if (lang === 'bash' || lang === 'sh' || lang === 'shell') {
        highlighted = highlightBash(line);
      } else if (lang === 'json') {
        highlighted = highlightJSON(line);
      }

      result += `  ${border} ${highlighted}\n`;
    }

    return result;
  }
}

/**
 * Basic JavaScript/TypeScript syntax highlighting
 */
function highlightJS(line: string): string {
  let result = line;

  // Keywords
  const keywords =
    /\b(const|let|var|function|return|if|else|for|while|class|interface|type|import|export|from|async|await|new|this|extends|implements)\b/g;
  result = result.replace(keywords, pc.magenta('$1'));

  // Strings
  result = result.replace(/(['"`])(?:(?!\1)[^\\]|\\.)*\1/g, (match) => pc.green(match));

  // Comments
  result = result.replace(/(\/\/.*$)/gm, pc.dim('$1'));

  // Numbers
  result = result.replace(/\b(\d+(?:\.\d+)?)\b/g, pc.yellow('$1'));

  return result;
}

/**
 * Basic bash syntax highlighting
 */
function highlightBash(line: string): string {
  let result = line;

  // Comments
  if (result.trim().startsWith('#')) {
    return pc.dim(result);
  }

  // Commands at start
  result = result.replace(/^(\s*)(\w+)/, (_, space, cmd) => space + pc.cyan(cmd));

  // Flags
  result = result.replace(/(\s)(--?[\w-]+)/g, (_, space, flag) => space + pc.yellow(flag));

  // Strings
  result = result.replace(/(['"])(?:(?!\1)[^\\]|\\.)*\1/g, (match) => pc.green(match));

  return result;
}

/**
 * Basic JSON syntax highlighting
 */
function highlightJSON(line: string): string {
  let result = line;

  // Keys
  result = result.replace(/"([^"]+)":/g, pc.cyan('"$1"') + ':');

  // String values
  result = result.replace(/:\s*"([^"]*)"/g, ': ' + pc.green('"$1"'));

  // Numbers and booleans
  result = result.replace(/:\s*(true|false|null|\d+(?:\.\d+)?)/g, (_, val) =>
    ': ' + pc.yellow(val)
  );

  return result;
}

/**
 * Simple function to render markdown without streaming
 */
export function renderMarkdown(text: string): string {
  const renderer = new MarkdownRenderer();
  renderer.push(text);
  return renderer.flush();
}

File: tui/src/render/markdown.zig
const std = @import("std");
const vaxis = @import("vaxis");
const syntax = @import("syntax.zig");

pub const MarkdownRenderer = struct {
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator) MarkdownRenderer {
        return .{ .allocator = allocator };
    }

    /// Render markdown to styled segments
    pub fn render(self: *MarkdownRenderer, markdown: []const u8) ![]Segment {
        var segments = std.ArrayList(Segment).init(self.allocator);
        var state = ParserState.init(self.allocator);
        defer state.deinit();

        var lines = std.mem.splitScalar(u8, markdown, '\n');
        var first = true;
        while (lines.next()) |line| {
            if (!first) {
                try segments.append(.{ .text = "\n", .style = .{} });
            }
            first = false;
            try self.parseLine(&segments, line, &state);
        }

        return segments.toOwnedSlice();
    }

    fn parseLine(self: *MarkdownRenderer, segments: *std.ArrayList(Segment), line: []const u8, state: *ParserState) !void {
        // Handle code blocks
        if (std.mem.startsWith(u8, line, "```")) {
            if (state.in_code_block) {
                // Closing code block - render with syntax highlighting
                try self.renderCodeBlock(segments, state);
                state.code_buffer.clearRetainingCapacity();
            } else {
                // Opening code block
                if (line.len > 3) {
                    state.code_language = line[3..];
                } else {
                    state.code_language = null;
                }
            }
            state.in_code_block = !state.in_code_block;
            return;
        }

        if (state.in_code_block) {
            // Accumulate code lines
            if (state.code_buffer.items.len > 0) {
                try state.code_buffer.append('\n');
            }
            try state.code_buffer.appendSlice(line);
            return;
        }

        // Headers
        if (line.len > 0 and line[0] == '#') {
            const header = self.parseHeader(line);
            try segments.append(.{
                .text = try self.allocator.dupe(u8, header.text),
                .style = header.style,
            });
            return;
        }

        // Blockquotes
        if (std.mem.startsWith(u8, line, "> ")) {
            try segments.append(.{
                .text = "│ ",
                .style = .{ .fg = .{ .index = 8 } },
            });
            try self.parseInline(segments, line[2..]);
            return;
        }

        // Unordered lists
        if (std.mem.startsWith(u8, line, "- ") or std.mem.startsWith(u8, line, "* ")) {
            try segments.append(.{
                .text = "  • ",
                .style = .{ .fg = .{ .index = 8 } },
            });
            try self.parseInline(segments, line[2..]);
            return;
        }

        // Ordered lists
        if (self.isOrderedListItem(line)) |num_end| {
            try segments.append(.{
                .text = try std.fmt.allocPrint(self.allocator, "  {s} ", .{line[0..num_end]}),
                .style = .{ .fg = .{ .index = 8 } },
            });
            try self.parseInline(segments, line[num_end + 2 ..]);
            return;
        }

        // Horizontal rule
        if (line.len >= 3 and (std.mem.eql(u8, line, "---") or std.mem.eql(u8, line, "***") or std.mem.eql(u8, line, "___"))) {
            try segments.append(.{
                .text = "────────────────────────────────────────",
                .style = .{ .fg = .{ .index = 8 } },
            });
            return;
        }

        // Regular line with inline formatting
        try self.parseInline(segments, line);
    }

    fn renderCodeBlock(self: *MarkdownRenderer, segments: *std.ArrayList(Segment), state: *ParserState) !void {
        if (state.code_buffer.items.len == 0) return;

        // Detect language
        const lang = if (state.code_language) |lang_str|
            syntax.detectLanguage(std.mem.trim(u8, lang_str, " \t\r\n"))
        else
            syntax.Language.unknown;

        // Highlight code
        const highlighted = try syntax.highlight(self.allocator, state.code_buffer.items, lang);
        defer highlighted.deinit();

        // Split by newlines and render each line with indentation
        var line_start: usize = 0;
        for (highlighted.items, 0..) |seg, i| {
            // Add indentation at start of each line
            if (line_start == 0 or (i > 0 and std.mem.endsWith(u8, highlighted.items[i - 1].text, "\n"))) {
                try segments.append(.{
                    .text = "  ",
                    .style = .{},
                });
            }

            // Add the styled segment
            try segments.append(.{
                .text = try self.allocator.dupe(u8, seg.text),
                .style = .{ .fg = .{ .index = seg.color } },
            });

            if (std.mem.endsWith(u8, seg.text, "\n")) {
                line_start = i + 1;
            }
        }
    }

    fn parseHeader(self: *MarkdownRenderer, line: []const u8) HeaderResult {
        _ = self;
        var level: u8 = 0;
        for (line) |char| {
            if (char == '#') level += 1 else break;
        }

        const text = std.mem.trim(u8, line[level..], " ");

        const style: vaxis.Cell.Style = switch (level) {
            1 => .{ .fg = .{ .index = 14 }, .bold = true, .ul_style = .single },
            2 => .{ .fg = .{ .index = 14 }, .bold = true },
            else => .{ .bold = true },
        };

        return .{ .text = text, .style = style };
    }

    const HeaderResult = struct {
        text: []const u8,
        style: vaxis.Cell.Style,
    };

    fn parseInline(self: *MarkdownRenderer, segments: *std.ArrayList(Segment), text: []const u8) !void {
        if (text.len == 0) return;

        var i: usize = 0;
        var current_start: usize = 0;

        while (i < text.len) {
            // Bold **text**
            if (i + 1 < text.len and text[i] == '*' and text[i + 1] == '*') {
                // Flush current text
                if (i > current_start) {
                    try segments.append(.{
                        .text = try self.allocator.dupe(u8, text[current_start..i]),
                        .style = .{},
                    });
                }

                // Find closing **
                const start = i + 2;
                i = start;
                while (i + 1 < text.len) : (i += 1) {
                    if (text[i] == '*' and text[i + 1] == '*') break;
                }

                if (i + 1 < text.len) {
                    try segments.append(.{
                        .text = try self.allocator.dupe(u8, text[start..i]),
                        .style = .{ .bold = true },
                    });
                    i += 2;
                } else {
                    // No closing found, treat as literal
                    try segments.append(.{
                        .text = "**",
                        .style = .{},
                    });
                    i = start;
                }
                current_start = i;
                continue;
            }

            // Italic *text* (but not **)
            if (text[i] == '*' and (i == 0 or text[i - 1] != '*') and (i + 1 >= text.len or text[i + 1] != '*')) {
                if (i > current_start) {
                    try segments.append(.{
                        .text = try self.allocator.dupe(u8, text[current_start..i]),
                        .style = .{},
                    });
                }

                const start = i + 1;
                i = start;
                while (i < text.len and text[i] != '*') : (i += 1) {}

                if (i < text.len) {
                    try segments.append(.{
                        .text = try self.allocator.dupe(u8, text[start..i]),
                        .style = .{ .italic = true },
                    });
                    i += 1;
                } else {
                    // No closing found, treat as literal
                    try segments.append(.{
                        .text = "*",
                        .style = .{},
                    });
                    i = start;
                }
                current_start = i;
                continue;
            }

            // Inline code `code`
            if (text[i] == '`') {
                if (i > current_start) {
                    try segments.append(.{
                        .text = try self.allocator.dupe(u8, text[current_start..i]),
                        .style = .{},
                    });
                }

                const start = i + 1;
                i = start;
                while (i < text.len and text[i] != '`') : (i += 1) {}

                if (i < text.len) {
                    try segments.append(.{
                        .text = try self.allocator.dupe(u8, text[start..i]),
                        .style = .{ .reverse = true },
                    });
                    i += 1;
                } else {
                    // No closing found, treat as literal
                    try segments.append(.{
                        .text = "`",
                        .style = .{},
                    });
                    i = start;
                }
                current_start = i;
                continue;
            }

            // Strikethrough ~~text~~
            if (i + 1 < text.len and text[i] == '~' and text[i + 1] == '~') {
                if (i > current_start) {
                    try segments.append(.{
                        .text = try self.allocator.dupe(u8, text[current_start..i]),
                        .style = .{},
                    });
                }

                const start = i + 2;
                i = start;
                while (i + 1 < text.len) : (i += 1) {
                    if (text[i] == '~' and text[i + 1] == '~') break;
                }

                if (i + 1 < text.len) {
                    try segments.append(.{
                        .text = try self.allocator.dupe(u8, text[start..i]),
                        .style = .{ .strikethrough = true },
                    });
                    i += 2;
                } else {
                    // No closing found, treat as literal
                    try segments.append(.{
                        .text = "~~",
                        .style = .{},
                    });
                    i = start;
                }
                current_start = i;
                continue;
            }

            // Links [text](url)
            if (text[i] == '[') {
                const link = self.parseLink(text[i..]);
                if (link) |l| {
                    if (i > current_start) {
                        try segments.append(.{
                            .text = try self.allocator.dupe(u8, text[current_start..i]),
                            .style = .{},
                        });
                    }

                    try segments.append(.{
                        .text = try self.allocator.dupe(u8, l.text),
                        .style = .{ .fg = .{ .index = 12 }, .ul_style = .single },
                        .link = try self.allocator.dupe(u8, l.url),
                    });

                    i += l.consumed;
                    current_start = i;
                    continue;
                }
            }

            i += 1;
        }

        // Flush remaining text
        if (current_start < text.len) {
            try segments.append(.{
                .text = try self.allocator.dupe(u8, text[current_start..]),
                .style = .{},
            });
        }
    }

    fn parseLink(self: *MarkdownRenderer, text: []const u8) ?LinkResult {
        _ = self;
        if (text.len == 0 or text[0] != '[') return null;

        // Find ]
        var i: usize = 1;
        while (i < text.len and text[i] != ']') : (i += 1) {}
        if (i >= text.len) return null;

        const link_text = text[1..i];

        // Expect (
        i += 1;
        if (i >= text.len or text[i] != '(') return null;

        // Find )
        const url_start = i + 1;
        i = url_start;
        while (i < text.len and text[i] != ')') : (i += 1) {}
        if (i >= text.len) return null;

        const url = text[url_start..i];

        return .{
            .text = link_text,
            .url = url,
            .consumed = i + 1,
        };
    }

    const LinkResult = struct {
        text: []const u8,
        url: []const u8,
        consumed: usize,
    };

    fn isOrderedListItem(self: *MarkdownRenderer, line: []const u8) ?usize {
        _ = self;
        var i: usize = 0;
        while (i < line.len and line[i] >= '0' and line[i] <= '9') : (i += 1) {}
        if (i > 0 and i + 1 < line.len and line[i] == '.' and line[i + 1] == ' ') {
            return i;
        }
        return null;
    }
};

pub const Segment = struct {
    text: []const u8,
    style: vaxis.Cell.Style = .{},
    link: ?[]const u8 = null,
};

const ParserState = struct {
    in_code_block: bool = false,
    code_language: ?[]const u8 = null,
    code_buffer: std.ArrayList(u8) = undefined,
    code_start_index: usize = 0,

    pub fn init(allocator: std.mem.Allocator) ParserState {
        return .{
            .code_buffer = std.ArrayList(u8).init(allocator),
        };
    }

    pub fn deinit(self: *ParserState) void {
        self.code_buffer.deinit();
    }
};

// Tests
test "parse headers" {
    const testing = std.testing;
    var renderer = MarkdownRenderer.init(testing.allocator);

    const segments = try renderer.render("# H1");
    defer testing.allocator.free(segments);

    try testing.expect(segments.len > 0);
    try testing.expect(segments[0].style.bold);
    try testing.expect(segments[0].style.fg.?.index == 14);
}

test "parse bold" {
    const testing = std.testing;
    var renderer = MarkdownRenderer.init(testing.allocator);

    const segments = try renderer.render("This is **bold** text");
    defer testing.allocator.free(segments);

    // Find the bold segment
    var found_bold = false;
    for (segments) |seg| {
        if (seg.style.bold and std.mem.eql(u8, seg.text, "bold")) {
            found_bold = true;
            break;
        }
    }
    try testing.expect(found_bold);
}

test "parse italic" {
    const testing = std.testing;
    var renderer = MarkdownRenderer.init(testing.allocator);

    const segments = try renderer.render("This is *italic* text");
    defer testing.allocator.free(segments);

    var found_italic = false;
    for (segments) |seg| {
        if (seg.style.italic and std.mem.eql(u8, seg.text, "italic")) {
            found_italic = true;
            break;
        }
    }
    try testing.expect(found_italic);
}

test "parse code block" {
    const testing = std.testing;
    var renderer = MarkdownRenderer.init(testing.allocator);

    const segments = try renderer.render(
        \\```zig
        \\const x = 10;
        \\```
    );
    defer testing.allocator.free(segments);

    // Should have code line
    var found_code = false;
    for (segments) |seg| {
        if (seg.style.dim and std.mem.indexOf(u8, seg.text, "const x = 10;") != null) {
            found_code = true;
            break;
        }
    }
    try testing.expect(found_code);
}

test "parse inline code" {
    const testing = std.testing;
    var renderer = MarkdownRenderer.init(testing.allocator);

    const segments = try renderer.render("Use `const` for constants");
    defer testing.allocator.free(segments);

    var found_code = false;
    for (segments) |seg| {
        if (seg.style.reverse and std.mem.eql(u8, seg.text, "const")) {
            found_code = true;
            break;
        }
    }
    try testing.expect(found_code);
}

test "parse list" {
    const testing = std.testing;
    var renderer = MarkdownRenderer.init(testing.allocator);

    const segments = try renderer.render("- Item 1");
    defer testing.allocator.free(segments);

    // Should have bullet point
    try testing.expect(segments.len > 0);
    try testing.expect(std.mem.indexOf(u8, segments[0].text, "•") != null);
}

test "parse numbered list" {
    const testing = std.testing;
    var renderer = MarkdownRenderer.init(testing.allocator);

    const segments = try renderer.render("1. First item");
    defer testing.allocator.free(segments);

    // Should have number
    try testing.expect(segments.len > 0);
    try testing.expect(std.mem.indexOf(u8, segments[0].text, "1") != null);
}

test "parse link" {
    const testing = std.testing;
    var renderer = MarkdownRenderer.init(testing.allocator);

    const segments = try renderer.render("[example](https://example.com)");
    defer testing.allocator.free(segments);

    var found_link = false;
    for (segments) |seg| {
        if (seg.link != null and std.mem.eql(u8, seg.text, "example")) {
            try testing.expectEqualStrings("https://example.com", seg.link.?);
            found_link = true;
            break;
        }
    }
    try testing.expect(found_link);
}

test "parse empty input" {
    const testing = std.testing;
    var renderer = MarkdownRenderer.init(testing.allocator);

    const segments = try renderer.render("");
    defer testing.allocator.free(segments);

    try testing.expect(segments.len == 0);
}

test "unclosed bold marker" {
    const testing = std.testing;
    var renderer = MarkdownRenderer.init(testing.allocator);

    const segments = try renderer.render("**unclosed");
    defer testing.allocator.free(segments);

    // Should treat ** as literal since no closing
    try testing.expect(segments.len > 0);
}
